Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go

```diff
@@ -0,0 +1,365 @@
+	return &connectionList{
+		conns:          []heldConn{},
+		maxSize:        maxSize,
+		tickSize:       tickSize,
+		timeSinceStart: 0,
+		timeout:        timeout,
+		svcName:        svcName,
+	}
+}
+
+func (l *connectionList) Add(conn net.Conn) {
+	if len(l.conns) >= l.maxSize {
+		// TODO: look for closed connections
+		glog.Errorf("max connections exceeded while waiting for idled service %s to awaken, dropping oldest", l.svcName)
+		var oldConn net.Conn
+		oldConn, l.conns = l.conns[0], l.conns[1:]
+		oldConn.Close()
+	}
+
+	l.conns = append(l.conns, heldConn{conn, l.timeSinceStart})
+}
+
+func (l *connectionList) Tick() {
+	l.timeSinceStart += l.tickSize
+	l.cleanOldConnections()
+}
+
+func (l *connectionList) cleanOldConnections() {
+	cleanInd := -1
+	for i, conn := range l.conns {
+		if l.timeSinceStart-conn.connectedAt < l.timeout {
+			cleanInd = i
+			break
+		}
+	}
+
+	if cleanInd > 0 {
+		oldConns := l.conns[:cleanInd]
+		l.conns = l.conns[cleanInd:]
+		glog.Errorf("timed out %v connections while waiting for idled service %s to awaken.", len(oldConns), l.svcName)
+
+		for _, conn := range oldConns {
+			conn.Close()
+		}
+	}
+}
+
+func (l *connectionList) GetConns() []net.Conn {
+	conns := make([]net.Conn, len(l.conns))
+	for i, conn := range l.conns {
+		conns[i] = conn.Conn
+	}
+	return conns
+}
+
+func (l *connectionList) Len() int {
+	return len(l.conns)
+}
+
+func (l *connectionList) Clear() {
+	for _, conn := range l.conns {
+		conn.Close()
+	}
+
+	l.conns = []heldConn{}
+}
+
+var (
+	needPodsWaitTimeout = 30 * time.Second
+	needPodsTickLen     = 5 * time.Second
+	maxHeldConnections  = 16
+)
+
+func newUnidlerSocket(protocol api.Protocol, ip net.IP, port int, eventRecorder record.EventRecorder) (proxySocket, error) {
+	host := ""
+	if ip != nil {
+		host = ip.String()
+	}
+
+	switch strings.ToUpper(string(protocol)) {
+	case "TCP":
+		listener, err := net.Listen("tcp", net.JoinHostPort(host, strconv.Itoa(port)))
+		if err != nil {
+			return nil, err
+		}
+		proxySocket := tcpProxySocket{Listener: listener, port: port}
+		return &tcpUnidlerSocket{tcpProxySocket: proxySocket, eventRecorder: eventRecorder}, nil
+	case "UDP":
+		addr, err := net.ResolveUDPAddr("udp", net.JoinHostPort(host, strconv.Itoa(port)))
+		if err != nil {
+			return nil, err
+		}
+		conn, err := net.ListenUDP("udp", addr)
+		if err != nil {
+			return nil, err
+		}
+		proxySocket := udpProxySocket{UDPConn: conn, port: port}
+		return &udpUnidlerSocket{udpProxySocket: proxySocket, eventRecorder: eventRecorder}, nil
+	}
+	return nil, fmt.Errorf("unknown protocol %q", protocol)
+}
+
+// tcpUnidlerSocket implements proxySocket.  Close() is implemented by net.Listener.  When Close() is called,
+// no new connections are allowed but existing connections are left untouched.
+type tcpUnidlerSocket struct {
+	tcpProxySocket
+	eventRecorder record.EventRecorder
```

Comment:
If you don't need all the methods on EventRecorder, make this be a simpler interface (with whichever method you need).